onnxruntime/core/providers/cpu/math/einsum_utils/einsum_auxiliary_ops.cc

```diff
@@ -272,6 +294,27 @@ bool IsTransposeRequired(size_t input_rank, const std::vector<size_t>& permutati
   return transpose_required;
 }
 
+// Reshape rather transpose. IsReshapeRequired determines whether or not it is possible.
+// The function creates a new tensor with a different shape but reuses the buffer coming from the input.
+std::unique_ptr<Tensor> Reshape(const Tensor& input, const std::vector<int64_t>& input_shape_override,
+                                const std::vector<size_t>& permutation, AllocatorPtr allocator) {
+  auto input_rank = input_shape_override.size();
+  ORT_ENFORCE(input_rank == permutation.size(), "Length of permutation must match the rank of the input to be permutated");
+
+  std::vector<int64_t> output_dims;
+  output_dims.reserve(input_rank);
+
+  for (const auto& dim : permutation) {
+    output_dims.push_back(input_shape_override[dim]);
+  }
+
+  // Pass in allocator as that will be used as an allocator deleter by the framework
+  // and it will de-allocate the memory for this intermediate tensor when it goes out of scope
+  std::unique_ptr<Tensor> output = onnxruntime::make_unique<Tensor>(input.DataType(), output_dims, (void*)input.DataRaw(), input.Location());
```

Comment:
A new tensor is created but it reuses the buffer of another one with a new shape or equal size. API of class Tensor does not seem to have a proper way of doing that.

Reply:
The Tensor class may need some enhancements to truly avoid this allocation (it may need to hang onto a shared_ptr) - we are creating a Tensor that doesn't own the memory - but the Tensor that actually owns the memory may be destructed and this will cause issues, won't it ?

While the generic case requires some thought, it might be okay if the input tensor to this method is actually the Einsum op's input, thus meaning the holding tensor is around until the op completes its execution and it may be okay to hang onto its buffer within other intermediate tensors within this operator and we are doing this as part of "pre-processing" the op's inputs - so as is it probably is fine but we should have some sort of asserts somewhere to ensure that the input tensor is truly the op's input and no other intermediate tensors...

Reply:
Yes, it would cause an issue. That's why I highlighted this part of the PR because I know it could cause an issue. However, in that case, everything happens in EinsteinOp, so the Tensor which owns the memory stays alive as long as it is needed. That's why I did it.

Reply:
What guarantees the input tensor stays alive? As this get called in a loop by EinsumTypedComputeProcessor<T>::Run can't the input go out of scope on the next iteration if it was the result from the previous iteration ?

e.g. first iteration in Run that calls PairwiseOperandProcess creates a new tensor which is saved as 'result'. next iteration passes this in as input. if we do the reshape on this iteration, the returned tensor replaces 'result'. at that point doesn't the buffer get freed as part of the assignment of the new result?

By chance the memory may still be accessible, but I'm not sure that the ownership semantics are correct.